### PR TITLE
fix(pricing_rule): transaction rule doesn't got apply on frontend

### PR DESCRIPTION
**Pull request by pipech**
## Fix for > [Bug] V12 - Pricing rule: Transaction rule doesn't got apply on frontend #22159

Apply transaction rule on frontend through `transaction.js` and `point_of_sale.js`.

This will work on both Sales Invoice page and Online POS page and both product discount and price discount.

### CODE REVIEW

File: erpnext/public/js/controllers/transaction.js [hunk 1]
```diff
@@ -569,6 +569,9 @@ erpnext.TransactionController = erpnext.taxes_and_totals.extend({
 										let key = item.name;
 										me.apply_rule_on_other_items({key: item});
 									}
+								},
+								() => {
+									me.frm.trigger('qty', cdt, cdn);
```
Comment (rohitwaghchaure) [inline erpnext/public/js/controllers/transaction.js]: Why trigger qty?

File: erpnext/public/js/controllers/transaction.js [hunk 2]
```diff
@@ -1264,6 +1267,43 @@ erpnext.TransactionController = erpnext.taxes_and_totals.extend({
 				}
 			}
 		});
+
+		const transaction_pricing_rule = this.frm.call({
```
Comment (rohitwaghchaure) [inline erpnext/public/js/controllers/transaction.js:1271]: This should be call in the callback of apply_pricing_rule

File: erpnext/public/js/controllers/transaction.js [hunk 3]
```diff
@@ -1264,6 +1267,43 @@ erpnext.TransactionController = erpnext.taxes_and_totals.extend({
...
+			method: "erpnext.accounts.doctype.pricing_rule.pricing_rule.apply_transaction_pricing_rule",
+			args: {	doc: me.frm.doc },
+			callback: function(r) {
+				if (!r.exc && r.message) {
+					r = r.message;
+					if ('price' in r) {
```
Comment (rohitwaghchaure) [inline erpnext/public/js/controllers/transaction.js:1277]: why not checking the value of the field price_or_product_discount

File: erpnext/public/js/controllers/transaction.js (hunk 1 above)
Comment (pipech) [reply erpnext/public/js/controllers/transaction.js]: To trigger apply_pricing_rule, actually i should change this to `this.apply_pricing_rule(item, true);`

File: erpnext/public/js/controllers/transaction.js (hunk 2 above)
Comment (pipech) [reply erpnext/public/js/controllers/transaction.js:1271]: May i ask why?

File: erpnext/public/js/controllers/transaction.js (hunk 3 above)
Comment (pipech) [reply erpnext/public/js/controllers/transaction.js:1277]: Return doesn't have `apply_transaction_pricing_rule` field.

File: erpnext/public/js/controllers/transaction.js (hunk 1 above)
Comment (pipech) [reply erpnext/public/js/controllers/transaction.js]: https://github.com/frappe/erpnext/blob/bc4737a1874b1c021c98fb50ae0c4b2dfecb5510/erpnext/public/js/controllers/transaction.js#L573-L575

**Comment by pipech** [comment]
@rohitwaghchaure sorry to tag you but could you please review this ?

File: erpnext/accounts/doctype/pricing_rule/pricing_rule.py [hunk 4]
```diff
@@ -175,6 +175,51 @@ def apply_pricing_rule(args, doc=None):
...
 	return out
 
+@frappe.whitelist()
+def apply_transaction_pricing_rule(doc):
+	from erpnext.accounts.doctype.pricing_rule.utils import apply_pricing_rule_on_transaction
+
+	# converting dict into doc
+	if isinstance(doc, string_types):
+		doc = json.loads(doc)
+	if doc:
```
Comment (sagarvora) [inline erpnext/accounts/doctype/pricing_rule/pricing_rule.py:185]: I don't know if this condition is needed at all. And if it _is_ needed, what happens if it's falsy?

**Review by sagarvora — CHANGES_REQUESTED** [review]

**Comment by rohitwaghchaure** [comment]
@pipech I was doing the testing for the POS, and I have found one issue that while removing the main item from the cart the free items is still remained in the cart. Please check below GIF

pos_testing_issue

I am closing this PR, requesting you to kindly fix this issue and reopen the PR

**Comment by federicocalvo** [comment]
@rohitwaghchaure @sagarvora Hi there, this PR is the only one that comes close to a correction in order to be able to use **pricing rules** with the **transaction** condition.

As of today, it still does not work and this PR was not reopened or revised.

Can it at least be reopened? We might be able to see about continuing it with our team.

Thanks

**Comment by sagarvora** [comment]
👋🏼 @federicocalvo 

Can you please open another one on `develop`? Or do you intend to use this on the `v12` POS?

**Comment by federicocalvo** [comment]
Hi @sagarvora, yes of course, but in our case it would be excluding the POS (I think it is very complex) but in the transactions as Sale Invoice (it should refer to Sales Order and Cuotation as well).

Then someone could separately try to solve it in point_of_sale.js.

**Comment by sagarvora** [comment]
> Hi @sagarvora, yes of course, but in our case it would be excluding the POS (I think it is very complex) but in the transactions as Sale Invoice (it should refer to Sales Order and Cuotation as well).
> 
> Then someone could separately try to solve it in point_of_sale.js.

Great, tag / assign me. I'll review it.

**Pull request closed without merging**